packages/office-ui-fabric-react/src/components/DetailsList/DetailsRowFields.tsx

```diff
@@ -19,84 +20,166 @@ export interface IDetailsRowFieldsProps {
+  * in three phases:
+  * 2. OldContentOutgiong: Render with the old data and animate it sliding out
+  * 1. NewContentIncoming: Render the new data and slide it in
+  * 0. Rest: resting state, remove all animation classes
+  * Every render() call will move the rendering phase closer to the rest state.
+  * The phases go 2->1->0 with a delay inbetween each to allow the animation to
+  * complete. Each cell will also check to make sure it changed, if its contents
+  * did not change, then it will not do any animation.
+*/
+export const enum DetailsRowFieldsRenderingPhase {
+  OldContentOutgoing = 2,
+  NewContentIncoming = 1,
+  Rest = 0
 }
 
 export class DetailsRowFields extends BaseComponent<IDetailsRowFieldsProps, IDetailsRowFieldsState> {
+  private static _getCellContent(props: IDetailsRowFieldsProps): React.ReactNode[] {
+    const { columns, item, itemIndex, onRenderItemColumn, shimmer } = props;
+
+    return columns.map(column => {
+      let cellContent;
+
+      try {
+        const render = column.onRender || onRenderItemColumn;
+
+        cellContent =
+          render && !shimmer ? render(item, itemIndex, column) : DetailsRowFields._getCellText(item, column);
+      } catch (e) {
+        /* no-op */
+      }
+
+      return cellContent;
+    });
+  }
+
+  private static _getCellText(item: any, column: IColumn): void {
+    let value = item && column && column.fieldName ? item[column.fieldName] : '';
+
+    if (value === null || value === undefined) {
+      value = '';
+    }
+
+    return value;
+  }
+
   constructor(props: IDetailsRowFieldsProps) {
     super(props);
 
-    this.state = this._getState(props);
+    this.state = {
+      columns: { ...{}, ...props.columns },
+      cellContent: DetailsRowFields._getCellContent(props),
+      item: { ...{}, ...props.item }
+    };
   }
 
-  public componentWillReceiveProps(newProps: IDetailsRowFieldsProps): void {
-    this.setState(this._getState(newProps));
+  public componentWillReceiveProps(nextProps: IDetailsRowFieldsProps) {
+    this.setState({
+      columns: nextProps.columns,
+      cellContent: DetailsRowFields._getCellContent(nextProps),
+      item: nextProps.item,
+      // when first called after ctor, set it to 0, otherwise it's new data and set it to 2
+      renderingPhase:
+        typeof this.state.renderingPhase === 'undefined'
+          ? DetailsRowFieldsRenderingPhase.Rest
+          : DetailsRowFieldsRenderingPhase.OldContentOutgoing,
+      oldColumns: { ...{}, ...this.props.columns }, // get by value instead of by reference
+      oldCellContent: DetailsRowFields._getCellContent(this.props),
+      oldItem: { ...{}, ...this.props.item }
+    });
   }
 
   public render(): JSX.Element {
-    const { columns, columnStartIndex, shimmer } = this.props;
-    const { cellContent } = this.state;
+    const { item, columns } = this.props;
+    const { cellContent, renderingPhase, oldColumns, oldCellContent, oldItem } = this.state;
+
+    if (renderingPhase) {
```

Comment:
The `render` call itself should not call `setState`. You should do this in `componentDidUpdate`